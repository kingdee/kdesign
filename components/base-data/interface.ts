import { tuple } from '../_utils/type'

export const Modes = tuple('multiple', 'single')
export type Mode = typeof Modes[number]

export type SelectValue = string | string[] | number | number[] | IOptionsProps[] | null

export interface IAdvancedSelectorProps<T = SelectValue> {
  prefixCls?: string // 样式前缀
  className?: string // 样式名
  children?: React.ReactNode
  style?: React.CSSProperties // 内联样式
  disabled?: boolean // 禁用
  dropdownStyle?: React.CSSProperties // 下拉样式
  notFoundContent?: React.ReactNode // 列表为空时文本
  placeholder?: string // 选择框默认文字
  showFrequent?: boolean // 是否显示历史记录与我的收藏面板
  showCollectIcon?: boolean // 是否显示收藏按钮
  showDetailIcon?: boolean // 是否显示详情按钮
  suffixIcon?: React.ReactNode // 选择框后缀图标
  mode?: Mode // 模式
  value?: IOptionsProps[] | IOptionsProps // 指定选中的条目
  columns?: [] // 下拉列表的展示列
  options?: [] // 下拉列表的数据
  size?: number // 下拉列表显示数量
  optionLabelProp?: string // 回填到选择框的属性
  dropdownFooterRender?: React.ReactNode // 自定义下拉框footer内容
  visible?: boolean
  collectList?: [] // 收藏列表
  historyList?: [] // 我的历史
  loading?: boolean // 加载中状态
  searchField?: string // 搜索字段 默认 all（全部字段）
  onChange?: () => void // 选中option调用此函数
  onSearch?: (val: string) => void // input的value变化时调用
  onSelect?: (value: T extends (infer I)[] ? I : T, option: React.ReactElement<any>) => void
  onShowDetail?: (value: IOptionsProps[]) => void
  onShowMore?: () => void
  onCollect?: (flag: boolean, value: IOptionsProps) => void
}

export interface IIndxPosListProps {
  end: number
  start: number
}

export interface IColumnsProps {
  title?: string
  key: string
}

export interface IOptionsProps {
  value?: string | number
  label?: string
}

export interface ISearchInfoProps {
  searchValue?: string
  previousInputValue: string
  previousEditValue?: string
  searchIndex?: number
  selectionStart?: number
  deleteEndIndx: number
  editOptions?: any[]
}
