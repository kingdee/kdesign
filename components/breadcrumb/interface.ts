import { ReactNode } from 'react'
import { DropDownProps } from '../dropdown'
import { tuple } from '../_utils/type'

export const ColorModelTypes = tuple('emphasize', 'weaken')
export type ColorModelType = typeof ColorModelTypes[number]

export interface IBreadcrumbProps {
  style?: Record<string, unknown> // 内联样式
  className?: string // 类名
  colorModel?: ColorModelType // 颜色模式，默认是 'emphasize'
  prefixCls?: string // 样式前缀
  items: Array<IBreadcrumbItem> // 面包屑导航的各项，支持传入 ReactNode 类型
  separator?: ReactNode // 分隔符，默认是 '/'
  children?: ReactNode // 子元素
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItems {
  item: IBreadcrumbItem
  index: number
  separator?: ReactNode
  openEllipsis?: boolean
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItem {
  className?: string // 类名
  title?: ReactNode
  dropdownProps?: DropDownProps
  href?: string // 链接地址，不能和path共用
  path?: string // 拼接地址，不能和href共用
  icon?: ReactNode
}

export interface IItemsWidth {
  width: number
  index: number
}
