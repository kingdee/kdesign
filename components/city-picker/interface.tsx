import * as React from 'react'
import { tuple } from '../_utils/type'

export type City = {
  id?: string | number
  country?: string
  province?: string
  name?: string
  [key: string]: any
}

export const BorderTypes = tuple('none', 'underline', 'bordered')
export type BorderType = typeof BorderTypes[number]

export const Sizes = tuple('large', 'middle', 'small') // 选择框大小
export type Size = typeof Sizes[number]

export type Type = 'domestic' | 'foreign'

export type CityList = Array<City>

export interface CityPickerProps {
  prefixCls?: string // 样式前缀
  size?: Size // 尺寸
  borderType?: BorderType // 边框类型
  notFoundContent?: React.ReactNode // 列表为空时文本
  showArrow?: boolean // 显示下拉小箭头
  defaultOpen?: boolean // 默认是否展开下拉列表
  allowClear?: boolean // 允许清除内容
  placeholder?: React.ReactNode // 选择框默认文字
  type?: Type
  loading?: boolean
  domesticList?: CityList
  foreignList?: CityList
  commonList?: CityList
  className?: string
  disabled?: boolean
  style?: React.CSSProperties
  title?: string
  clearIcon?: React.ReactNode // 清除图标
  suffixIcon?: React.ReactNode // 选择框后缀图标
  optionHighlightProps?: string // 搜索时 根据该属性值高亮城市名
  value?: string | number
  defaultValue?: string | number
  description?: React.ReactNode
  showDescription?: boolean
  onChange?: (value: string | number, city: City) => void
  onBlur?: (e: React.ChangeEvent<HTMLSpanElement>) => void
  onFocus?: () => void
  onSearch?: (value: string) => void
  onClear?: () => void
  itemRender?: (city: City) => React.ReactNode
  onTabPaneChange?: (id: Type) => void
}

export interface ICityPickerOptionProps {
  disabled?: boolean
  city?: City
  value?: string | number
  children?: React.ReactNode
  className?: string
  style?: React.CSSProperties
  onChangeSelect?: (city: City | null) => void
  renderCityInfo?: (city: City, flag?: boolean, symbol?: string) => React.ReactNode
  itemRender?: (city: City) => React.ReactNode
}
