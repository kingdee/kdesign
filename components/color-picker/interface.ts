import { tuple } from '../_utils/type'
import { PopperProps } from '../_utils/usePopper'

export const ColorTypes = tuple('HEX', 'HSB', 'RGB', 'HSL', 'themeColor')
export type ColorType = typeof ColorTypes[number]
export interface IColorTypesObj {
  type: ColorType
  value: string
}

export const BorderTypes = tuple('underline', 'bordered')
export type BorderType = typeof BorderTypes[number]
export interface IColorPickerProps extends PopperProps {
  // 设置颜色选择器的值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  value: string
  // 设置颜色选择器的默认值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  defaultValue: string
  // 选择器输入框类名
  className: string
  // 选择输入框样式
  style: React.CSSProperties
  // 输入框类型
  borderType: BorderType
  // 输入框右侧的图标
  suffixIcon: (rgbColor: string) => React.ReactNode
  // 功能色名称
  functionalColorName: string
  // 功能色色值
  functionalColor: string
  // 功能色开关的名称
  switchName: { name: string; internationalName?: string }
  // 是否跟随功能色,定义了功能色才能控制显示跟随主题色开关
  showSwitch: boolean
  // 是否展示色彩格式转换框
  showColorTransfer: boolean
  // 是否展示系统预设颜色
  showPresetColor: boolean
  // 占位字符
  placeholder: string
  // 用户自定义的预设颜色
  presetColor: string[]
  // 是否展示拾色容器
  showColorPickerBox: { showBox: boolean; showHue: boolean; showOpacity: boolean }
  // 颜色选择面板显隐配置
  defaultOpen: boolean
  // 当颜色选择器的值发生变化时触发的回调函数
  onChange: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}

export interface IColorPickerPanelProps {
  alpha: number
  alphaNoVerifyVal: string
  clickedColorIndex?: number
  colTypeArr: Array<IColorTypesObj>
  correctColorValue: string
  currentColorType: IColorTypesObj['type']
  functionalColor?: string
  functionalColorName: string
  switchName: { name: string; internationalName?: string }
  inputRef: any
  isFollow: boolean
  presetColor?: string[]
  showSwitch?: boolean
  showColorTransfer?: boolean
  showPresetColor?: boolean
  showColorPickerBox?: { showBox?: boolean; showHue?: boolean; showOpacity?: boolean }
  showColorPickerPanel: boolean
  value: string
  setCurrentColorType: (currentColorType: IColorTypesObj['type']) => void
  setColTypeArr: (colTypeArr: Array<IColorTypesObj>) => void
  setClickedColorIndex: (clickedColorIndex: number) => void
  setIsFollow: (isFollow: boolean) => void
  setAlphaNoVerifyVal: (NoVerifyVal: string) => void
  setAlpha: (alpha: number) => void
  setCorrectColorValue: (colorValue: string) => void
  setInputColorValue: (colorValue: string) => void
  setShowPanel: (showPanel: boolean) => void
  onChange?: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}
