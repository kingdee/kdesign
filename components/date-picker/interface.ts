import { BorderType, InputSiteType } from '../input/input'
import { DisabledDataProps } from './date-picker'

export type WeekTitleType = [string, string, string, string, string, string, string]
export type MonthTitleType = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
]

export interface Locale {
  // locale: string
  'DatePicker.placeholder': string
  'DatePicker.yearPlaceholder': string
  'DatePicker.quarterPlaceholder': string
  'DatePicker.monthPlaceholder': string
  'DatePicker.weekPlaceholder': string
  'DatePicker.timePlaceholder': string
  'DatePicker.now': string
  'DatePicker.confrim': string
  'DatePicker.today': string
  'DatePicker.year': string
  'DatePicker.month': string
  'DatePicker.weekTitle': WeekTitleType
  'DatePicker.monthTitle': MonthTitleType
  'DatePicker.rangePlaceholder': [string, string]
  'DatePicker.rangeYearPlaceholder': [string, string]
  'DatePicker.rangeMonthPlaceholder': [string, string]
  'DatePicker.rangeWeekPlaceholder': [string, string]
  'DatePicker.rangeQuarterPlaceholder': [string, string]
  'DatePicker.rangeTimePlaceholder': [string, string]
}

export interface InnerLocale {
  locale: string
  placeholder: string
  yearPlaceholder: string
  quarterPlaceholder: string
  monthPlaceholder: string
  weekPlaceholder: string
  timePlaceholder: string
  now: string
  confrim: string
  today: string
  year: string
  month: string
  weekTitle: WeekTitleType
  monthTitle: MonthTitleType
  rangePlaceholder: [string, string]
  rangeYearPlaceholder: [string, string]
  rangeMonthPlaceholder: [string, string]
  rangeWeekPlaceholder: [string, string]
  rangeQuarterPlaceholder: [string, string]
  rangeTimePlaceholder: [string, string]
}

export type InnerLocaleKey = keyof InnerLocale

export type PanelMode = 'time' | 'date' | 'week' | 'month' | 'quarter' | 'year' | 'decade'

export type PickerMode = Exclude<PanelMode, 'datetime' | 'decade'>

export type DateType = Date

export type CustomFormat = (value: DateType) => string

export type DateFormat = string

export type DateFormats = Array<string>

export type NullableDateType = DateType | null | undefined

export type OnSelect = (value: DateType, type: 'key' | 'mouse' | 'submit') => void
export interface DisabledTimes {
  disabledHours?: () => number[]
  disabledMinutes?: (hour: number) => number[]
  disabledSeconds?: (hour: number, minute: number) => number[]
}

export type DisabledTime = (date: DateType | null) => DisabledTimes

export type OnPanelChange = (value: DateType, mode: PanelMode) => void

export type EventValue = DateType | null
export type RangeValue = [EventValue, EventValue] | null

export type RangeList = {
  label: string
  onClick: () => void
  onMouseEnter: () => void
  onMouseLeave: () => void
}[]

export interface Components {
  button?: React.ComponentType | string
  rangeItem?: React.ComponentType | string
}

export interface PanelSharedProps {
  value?: NullableDateType
  defaultPickerValue?: DateType
  locale: InnerLocale
  disabledDate?: DisabledDataProps

  prevIcon?: React.ReactNode
  nextIcon?: React.ReactNode
  superPrevIcon?: React.ReactNode
  superNextIcon?: React.ReactNode

  // 面板操作（焦点操作、键盘操作）next version
  // operationRef: React.MutableRefObject<PanelRefProps>

  // onPanelChange: (mode: PanelMode | null, viewValue: DateType) => void
}

export type DateRender = (currentDate: DateType, today: DateType) => React.ReactNode

export type CellRender = (currentDate: DateType) => React.ReactNode

export type RangeType = 'start' | 'end'

export interface RangeInfo {
  range: RangeType
}

export type RangeDateRender = (currentDate: DateType, today: DateType, info: RangeInfo) => React.ReactNode

export interface SharedTimeProps extends DisabledTimes {
  format?: string
  showNow?: boolean
  showHour?: boolean
  showMinute?: boolean
  showSecond?: boolean
  use12Hours?: boolean
  hourStep?: number
  minuteStep?: number
  secondStep?: number
  hideDisabledOptions?: boolean
  defaultValue?: DateType
}

export interface OutInputCommonProps {
  format: DateFormat
  picker: PickerMode
  id?: string
  autoFocus?: boolean
  allowClear?: boolean
  className?: string
  borderType?: BorderType
  style?: React.CSSProperties
  size?: InputSiteType
  readOnly?: boolean
  suffixIcon?: React.ReactNode
  clearIcon?: React.ReactNode

  onFocus?: React.FocusEventHandler<HTMLInputElement>
  onBlur?: React.FocusEventHandler<HTMLInputElement>
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>
  onClick?: React.MouseEventHandler<HTMLDivElement>
  onContextMenu?: React.MouseEventHandler<HTMLDivElement>
}

export interface InputCommonProps {
  panelDivRef: React.RefObject<HTMLDivElement>
  needConfirmButton: boolean
  open: boolean
  prefixCls: string
  locale: InnerLocale
  dataOrAriaProps?: Record<string, any>
  status?: 'error'
}

export interface OutPopperProps {
  dropdownClassName?: string
  popupStyle?: React.CSSProperties
  showArrow?: boolean
}

export interface TimeUnit {
  label: React.ReactText
  value: number
  disabled: boolean
}
