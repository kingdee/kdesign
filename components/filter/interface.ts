import React from 'react'
import { Menu } from '../dropdown/menu'
import { IQuickSearchProps } from '../search'
import { DirectionType } from '../config-provider/ConfigContext'

type TabKey = 'condition' | 'scheme'

export type TFilterValue = {
  [key: string]: string[] | IScheme
}

export type TOption = {
  value?: string
  label?: string
  component?: React.ReactNode
}

export interface ICondition {
  key: string
  label: string
  options: TOption[]
  required?: boolean
  multiple?: boolean
  addition?: React.ReactNode
}

export interface IField {
  name: string
  label: string
  required?: boolean
  operators: React.ReactElement
  assignOptions: Record<string, React.ReactElement>
}

export interface ISchemeItem {
  fieldName: string
  value: string
  operator: string
  required?: boolean
}

export interface IScheme {
  name?: string
  preset?: boolean
  items: ISchemeItem[]
}

export interface IFilterProps {
  spread?: boolean
  fields?: IField[]
  className?: string
  schemes?: IScheme[]
  value?: TFilterValue
  defaultTabKey?: TabKey
  defaultSpread?: boolean
  title?: React.ReactNode
  conditions?: ICondition[]
  search?: IQuickSearchProps
  style?: React.CSSProperties
  defaultValue?: TFilterValue
  onSchemeSave?: (scheme: IScheme) => void
  onSpreadChange?: (spread: boolean) => void
  getSchemeSettingMenu?: (scheme: IScheme) => typeof Menu
  onChange?: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface IConditionProps {
  tab: string
  prefixCls: string
  value: TFilterValue
  conditions: ICondition[]
  FilterLangMsg: Record<string, any>
  onConditionRemove: (key: string, e: React.MouseEvent) => void
  onConditionChange: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface ISchemeProps {
  tab: string
  fields: IField[]
  prefixCls: string
  schemes: IScheme[]
  direction: DirectionType
  value: { scheme: IScheme }
  FilterLangMsg: Record<string, any>
  onSchemeSave: (scheme: IScheme) => void
  onSchemeSearch: (scheme: IScheme) => void
  getSchemeSettingMenu?: (scheme: IScheme) => void
}
