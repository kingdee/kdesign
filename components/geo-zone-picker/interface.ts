import { tuple } from '../_utils/type'
export const BorderTypes = tuple('underline', 'bordered')
export type BorderType = typeof BorderTypes[number]
const TabTypes = tuple('PROVINCE', 'CITY', 'DISTRICT')
export type TabType = typeof TabTypes[number]
export interface Country {
  code: number | string
  name: string
}

export interface Region {
  code: string
  name: string
  group?: string
  children?: Region[]
}

// 声明数组对象的类型
export type RegionArray = Array<Region>

export interface RegionValue {
  countryCode?: string | number
  countryName?: string
  provinceCode?: string | number
  provinceName?: string
  cityCode?: string | number
  cityName?: string
  districtCode?: string | number
  districtName?: string
}
