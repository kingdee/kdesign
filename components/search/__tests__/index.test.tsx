import React from 'react'
import { render, mount } from 'enzyme'
import Search from '../index'
import { SearchSizeTypes, BorderTypes } from '../interface'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'

describe('Search', () => {
  // 1.mount test
  mountTest(Search)

  // 2.render test
  it('renders correctly', () => {
    SearchSizeTypes.forEach((size) => {
      expect(render(<Search size={size} />)).toMatchSnapshot()
    })

    BorderTypes.forEach((type) => {
      expect(render(<Search borderType={type} />)).toMatchSnapshot()
    })

    const wrapper = render(
      <Search>
        <Search.Option value="all" tag="全部" />
      </Search>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  // 3. warns in component
  describe('3. warns in component', () => {
    it('warns if type is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const size = 'size-test'
      render(<Search size={size as any} />)
      expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-search: cannot found search size 'size-test'")
    })
  })

  // 4. render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Search>
          {null}
          {undefined}
        </Search>,
      ),
    ).toMatchSnapshot()
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Search></Search>)
    expect((wrapper.type() as any).displayName).toBe('Search')
  })

  // 6. class state
  it('should class use right', () => {
    const wrapper = mount(
      <Search type="quick-search" className="my-test" style={{ color: 'red' }} disabled data-test="test"></Search>,
    )
    expect(wrapper.prop('data-test')).toEqual('test')
    expect(wrapper.find('.kd-quick-search')).toHaveStyle('color', 'red')
    expect(wrapper.find('.kd-quick-search')).toHaveClassName('.my-test')
    expect(wrapper.find('.kd-quick-search-disabled')).toExist()
  })

  // 7. component interaction(event)
  // Focus
  it('Focus', () => {
    const onFocus = jest.fn()
    const wrapper = mount(<Search type="quick-search" onFocus={onFocus}></Search>)
    wrapper.find('.kd-quick-search-input').simulate('focus')
    expect(onFocus).toHaveBeenCalled()
    expect(wrapper.find('.kd-quick-search-focused').length).toBeGreaterThanOrEqual(1)
  })

  // Blur
  it('Blur', () => {
    const onBlur = jest.fn()
    const wrapper = mount(<Search type="quick-search" onBlur={onBlur}></Search>)
    wrapper.find('.kd-quick-search-input').simulate('blur')
    expect(onBlur).toHaveBeenCalled()
  })

  // dropDown desc
  it('should have default dropDown desc', () => {
    const wrapper = mount(<Search type="quick-search" />)
    wrapper.find('.kd-quick-search-input').simulate('focus')
    expect(wrapper.find('.kd-quick-search-focused').length).toBeGreaterThanOrEqual(1)
  })

  // placeholder
  it('should show placeholder', () => {
    const wrapper = mount(<Search type="quick-search" />)
    expect(wrapper.find('.kd-quick-search-selection-placeholder').length).toBeGreaterThanOrEqual(1)
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    expect(wrapper.find('.kd-quick-search-selection-placeholder').length).toBeGreaterThanOrEqual(0)
  })

  // notFoundContent
  it('should have default notFoundContent', () => {
    const wrapper = mount(<Search type="quick-search" />)
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    expect(wrapper.find('.kd-quick-search-dropdown-empty').length).toBeTruthy()
  })

  it('should support set notFoundContent to null', () => {
    const wrapper = mount(<Search type="quick-search" notFoundContent={null} />)
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    expect(wrapper.find('.kd-quick-search-dropdown-empty').length).toBeTruthy()
  })
  it('onChange', () => {
    let value = ''
    const onChange = jest.fn((val) => {
      value = val
    })
    const wrapper = mount(
      <Search type="quick-search" onChange={onChange}>
        <Search.Option value={1} tag="全部" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    wrapper.find('.kd-quick-search-option').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    expect(value).toStrictEqual([1])
  })

  it('onSelect', () => {
    let value = ''
    const onSelect = jest.fn((val) => {
      value = val
    })
    const wrapper = mount(
      <Search type="quick-search" onSelect={onSelect}>
        <Search.Option value={1} tag="全部" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    wrapper.find('.kd-quick-search-option').at(0).simulate('click')
    expect(onSelect).toHaveBeenCalled()
    expect(value).toBe(1)
  })

  // changes when multiple Select
  it('changes when multiple Select', () => {
    const onChange = jest.fn()
    const { Option } = Search
    const wrapper = mount(
      <Search type="quick-search" placeholder="请输入名称" onChange={onChange}>
        <Option value={1} tag="全部" />
        <Option value={2} tag="发现人" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '1' } })
    wrapper.find('.kd-quick-search-option').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    wrapper.find('.kd-quick-search-option').at(1).simulate('click')
    expect(onChange).toHaveBeenCalled()
  })

  // enter up down backspace
  it('keydown', () => {
    let value = ''
    const onSelect = jest.fn((val) => {
      value = val
    })
    const { Option } = Search
    const wrapper = mount(
      <Search type="quick-search" placeholder="请输入名称" onSelect={onSelect}>
        <Option value={1} tag="全部" />
        <Option value={2} tag="发现人" />
        <Option value={3} tag="处理人" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '1' } })
    wrapper.find('.kd-quick-search-input').simulate('keydown', { keyCode: 40 })
    wrapper.find('.kd-quick-search-input').simulate('keydown', { keyCode: 13 })
    expect(value).toBe(2)
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '1' } })
    wrapper.find('.kd-quick-search-input').simulate('keydown', { keyCode: 38 })
    wrapper.find('.kd-quick-search-input').simulate('keydown', { keyCode: 13 })
    expect(value).toBe(3)
    wrapper.find('.kd-quick-search-input').simulate('keydown', { keyCode: 8 })
    expect(wrapper.find('.kd-quick-search-selection-overflow-item').length).toBe(1)
  })

  it('delete tag when selected', () => {
    let value = ''
    const onChange = jest.fn((val) => {
      value = val
    })
    const { Option } = Search
    const wrapper = mount(
      <Search type="quick-search" placeholder="请输入名称" onChange={onChange}>
        <Option value={1} tag="全部" />
        <Option value={2} tag="发现人" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '1' } })
    wrapper.find('.kd-quick-search-option').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    expect(wrapper.find('.kd-quick-search-selection-overflow-item').length).toBe(1)
    wrapper.find('.kd-quick-search-selection-overflow-item-del').at(0).simulate('click')
    expect(onChange).toHaveBeenCalled()
    expect(value).toStrictEqual([])
  })

  // nlpSearch
  it('changes when set nlpSearch', () => {
    const onChange = jest.fn()
    const { Option } = Search
    const nlpSearch = {
      isSupportNlpSearch: true,
      nlpSearchLoading: true,
    }
    const wrapper = mount(
      <Search type="quick-search" placeholder="请输入名称" onChange={onChange} nlpSearch={nlpSearch}>
        <Option value={1} tag="全部" />
        <Option value={2} tag="发现人" />
      </Search>,
    )
    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '1' } })
    expect(wrapper.find('.kd-quick-search-npl').length).toBeTruthy()
  })

  it('prefix suffix', () => {
    const wrapper = mount(<Search prefix="prefix-test" suffix="suffix-test" />)
    expect(wrapper.find('.kd-input-prefix').text()).toBe('prefix-test')
    expect(wrapper.find('.kd-input-suffix').text()).toBe('suffix-test')
  })

  it('quick search api test', () => {
    const onSearch = jest.fn()
    const wrapper = mount(
      <Search
        type="quick-search"
        onSearch={onSearch}
        dropdownStyle={{ color: 'red' }}
        listHeight={100}
        desc={['1', '2']}
        tags={[
          { value: '1', tag: '全部' },
          { value: '2', tag: '发现人' },
        ]}
      />,
    )

    wrapper.find('.kd-quick-search-input').simulate('change', { target: { value: '12' } })
    expect(onSearch).toBeCalled()

    // dropdownStyle
    expect(wrapper.find('.kd-quick-search-dropdown')).toHaveStyle('color', 'red')

    // listHeight
    expect(wrapper.find('.kd-quick-search-dropdown-scroll')).toHaveStyle('maxHeight', 100)

    // tags
    expect(wrapper.find('.kd-quick-search-option').length).toBe(2)
  })

  // 8.config provider
  it('should config use config provider', () => {
    const localeData = {
      'Search.placeholder': 'Please enter what you want to search for',
    }
    const searchConfig = {
      compDefaultProps: {
        Search: {
          size: 'small',
        },
      },
      localeConfig: { localeData, locale: 'zh-EN' },
    }
    const wrapper = mount(
      <ConfigProvider value={searchConfig}>
        <Search />
      </ConfigProvider>,
    )
    expect(wrapper.find('.kd-search')).toHaveClassName('.kd-search-size-small')
    expect(wrapper.find('input').prop('placeholder')).toBe('Please enter what you want to search for')
  })

  // 9. ref test
  it('should get Demo element from ref', () => {
    const ref = React.createRef()
    mount(<Search ref={ref} type="quick-search" />)
    expect(ref.current instanceof HTMLElement).toBe(true)
    expect((ref.current as HTMLElement).classList.contains('kd-quick-search')).toBe(true)
  })
})
