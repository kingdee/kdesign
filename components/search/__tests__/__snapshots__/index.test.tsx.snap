// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search render no child without errors 1`] = `
<Search>
  <div
    className="kd-search kd-search-size-middle"
  >
    <Input
      borderType="none"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onPressEnter={[Function]}
      placeholder="请输入需要搜索的内容"
      prefix={null}
      size="middle"
      suffix={null}
    >
      <ClearableInput
        borderType="none"
        disabled={false}
        element={
          <input
            className="kd-input kd-input-size-middle kd-input-borderless"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="请输入需要搜索的内容"
            prefix={null}
            type="text"
            value=""
          />
        }
        focused={false}
        handleReset={[Function]}
        inputType="input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onPressEnter={[Function]}
        placeholder="请输入需要搜索的内容"
        prefix={null}
        prefixCls="kd-input"
        size="middle"
        suffix={null}
        type="text"
        value=""
      >
        <input
          className="kd-input kd-input-size-middle kd-input-borderless"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          placeholder="请输入需要搜索的内容"
          prefix={null}
          type="text"
          value=""
        />
      </ClearableInput>
    </Input>
  </div>
</Search>
`;

exports[`Search render null or undefined without errors 1`] = `
<Search>
  <div
    className="kd-search kd-search-size-middle"
  >
    <Input
      borderType="none"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onPressEnter={[Function]}
      placeholder="请输入需要搜索的内容"
      prefix={null}
      size="middle"
      suffix={null}
    >
      <ClearableInput
        borderType="none"
        disabled={false}
        element={
          <input
            className="kd-input kd-input-size-middle kd-input-borderless"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="请输入需要搜索的内容"
            prefix={null}
            type="text"
            value=""
          />
        }
        focused={false}
        handleReset={[Function]}
        inputType="input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onPressEnter={[Function]}
        placeholder="请输入需要搜索的内容"
        prefix={null}
        prefixCls="kd-input"
        size="middle"
        suffix={null}
        type="text"
        value=""
      >
        <input
          className="kd-input kd-input-size-middle kd-input-borderless"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          placeholder="请输入需要搜索的内容"
          prefix={null}
          type="text"
          value=""
        />
      </ClearableInput>
    </Input>
  </div>
</Search>
`;

exports[`Search renders correctly 1`] = `
<div
  class="kd-search kd-search-size-large"
>
  <input
    class="kd-input kd-input-size-large kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 2`] = `
<div
  class="kd-search kd-search-size-middle"
>
  <input
    class="kd-input kd-input-size-middle kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 3`] = `
<div
  class="kd-search kd-search-size-small"
>
  <input
    class="kd-input kd-input-size-small kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 4`] = `
<div
  class="kd-search kd-search-size-middle kd-search-borderless"
>
  <input
    class="kd-input kd-input-size-middle kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 5`] = `
<div
  class="kd-search kd-search-size-middle"
>
  <input
    class="kd-input kd-input-size-middle kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 6`] = `
<div
  class="kd-search kd-search-size-middle kd-search-border"
>
  <input
    class="kd-input kd-input-size-middle kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;

exports[`Search renders correctly 7`] = `
<div
  class="kd-search kd-search-size-middle"
>
  <input
    class="kd-input kd-input-size-middle kd-input-borderless"
    placeholder="请输入需要搜索的内容"
    type="text"
    value=""
  />
</div>
`;
